docs/develop-resources/api-conventions/validations.md

```diff
@@ -0,0 +1,158 @@
+
+
+```
+type PrivatePoolV1Config_NetworkConfigSpec struct {
+        // +kubebuilder:validation:XValidation:rule="self == oldSelf",message="the field is immutable"
+        // Immutable. The network definition that the workers are peered
+        //  to. If this section is left empty, the workers will be peered to
+        //  `WorkerPool.project_id` on the service producer network.
+        PeeredNetworkRef refv1beta1.ComputeNetworkRef `json:"peeredNetworkRef,omitempty"`
+}
+```
+
+Note: 
+
+
+* Kubernetes supports CEL since 1.26 which is the oldest GKE cluster version (written at ). Using CEL shall not cause GKE cluster version issues, but we shall reevaluate this when the condition changes.
+* Some ConfigConnector resources use webhook to validate immutable fields. We can continue using that, but for external contributors CEL is much easier for self-learning. 
+
+We only apply CEL on immutable fields.** More complicated CEL validation requires further discussion.**
+
+One future improvement can be having generate-types detect the “Immutable.” comment and add the kubebuilder tag automatically. 
+
+
+## Rule 3: Parent
+
+Parent field should be required and immutable.
+
+
+### Required
+
+Suggest using the following `inline` struct
+
+
+```
+type CloudBuildWorkerPoolSpec struct {
+  // +required   
+  Parent `json:",inline"`
+}
+
+type Parent struct {
+    // +required      
+    ProjectRef *refv1beta1.ProjectRef `json:"projectRef"`
+
+    // +required
+    Location string `json:"location"`
+}
+```
+
+
+### Immutable
+
+Since most parent is a reference to another resource which can either be `&lt;kind>Ref.external `or `&lt;kind>Ref.name` , the immutable validation needs to be done at the controller validation level. For example, switching `&lt;kind>Ref.external `and `&lt;kind>Ref.name `are allowed if referring to the same resource.
+
+See  for general reference guide;
+
+See  for parent validation with `status.externalRef`.   
+
+
+## Rule 4: No `anyOf`, `oneOf`, or` allOf`
+
+Short answer: Not now.
+
+Full answer: 
+
+OpenAPI supports `oneOf`, `anyOf` and `allOf`. Thus, in theory these should be the CRD level rules. The Direct Resource CRD is auto-generated by controller-gen, which only reads the kubebuilder tag rules. But the kubebuilder tags [do not (and most likely won’t)](https://github.com/kubernetes-sigs/controller-tools/issues/461) support `oneOf`, `anyOf` or `allOf`.
+
+To set this as CRD level rule, ConfigConnector shall build its own CRD transformer.  Ideally, this could be a ConfigConnector tag similar to kubebuilder, but this requires integrating with controller-gen which is a non-trivial amount of work. Other options like webhook, OpenAPI schema modifier are not self-explaining and easy enough to use, considering that Direct Resource shall be open to external contributors. Thus, ConfigConnector does not use CRD level validation for `oneOf`, `anyOf` and `allOf`.
```

Comment:
Does this exist yet?

Reply:
Not yet. We may want to build it someday. Defer to the requirements